tests/strictness-checks/Partial.hs

```diff
@@ -0,0 +1,170 @@
+  arbitrary = generatePartial arbitrary
+  shrink (Specification (Just value)) = fmap (Specification . Just) (shrink value)
+  shrink (Specification Nothing) = [ ]
+
+generatePartial :: Gen value -> Gen (Specification value)
+generatePartial generatorOfValues = frequency
+      [ (1, return (Specification Nothing))
+      , (9, fmap (Specification . Just) generatorOfValues) ]
+
+-- | Extract an error from a value that might be somewhere undefined.
+save :: NFData value => value -> IO (Either String value)
+save = fmap (first errorMessage) . try @ErrorCall . evaluate . force
+
+-- | Total values are returned as they are. They are never looked into.
+newtype Total value = Total {total :: value} deriving (Functor, Foldable, Traversable, Arbitrary)
+instance Show value => Show (Total value)
+  where
+    show = show . total
+
+type family MapPartial (typeConstructor :: * -> *) value where
+  MapPartial _ (Total value) = Total value
+  MapPartial typeConstructor (Specification value) = typeConstructor (MapPartial typeConstructor value)
+  MapPartial typeConstructor (trifunctor first second third) = trifunctor (MapPartial typeConstructor first) (MapPartial typeConstructor second) (MapPartial typeConstructor third)
+  MapPartial typeConstructor (bifunctor first second) = bifunctor (MapPartial typeConstructor first) (MapPartial typeConstructor second)
+  MapPartial typeConstructor (functor value) = functor (MapPartial typeConstructor value)
+  MapPartial _ anythingElse = anythingElse
+
+newtype Labelled value = Labelled {labelled :: Either [Natural] value} deriving (Functor, Foldable, Traversable)
+instance Show value => Show (Labelled value) where
+  show (Labelled (Right value)) = show value
+  show (Labelled (Left ordinals)) = "\"" ++ "#" ++ (List.intercalate "/" . reverse . fmap show) ordinals ++ "\""
+
+-- | Provide a discernible label for every value that is not defined.
+label :: Specification value -> Reader [Natural] (Labelled value)
+label = maybe (fmap (Labelled . Left) ask) (return . Labelled . Right) . partial
+
+class LabelRecursively partial
+  where
+    labelRecursively :: partial -> Reader [Natural ] (MapPartial Labelled partial)
+
+instance LabelRecursively (Total value)
+  where
+    labelRecursively = return
+
+instance LabelRecursively value => LabelRecursively (Specification value)
+  where
+    labelRecursively = (bind . traverse) labelRecursively . label
+
+instance (LabelRecursively a, LabelRecursively b, LabelRecursively c) => LabelRecursively (a, b, c)
+  where
+    labelRecursively (a, b, c) = pure (, , ) <*> local (0: ) (labelRecursively a) <*> local (1: ) (labelRecursively b) <*> local (2: ) (labelRecursively c)
+
+instance {-# overlappable #-}
+  ( Bitraversable bifunctor
+  , LabelRecursively firstValue
+  , LabelRecursively secondValue
+  , MapPartial Labelled (bifunctor firstValue secondValue) ~ bifunctor (MapPartial Labelled firstValue) (MapPartial Labelled secondValue)
```

Comment:
This equality seems always satisfied by definition of `MapPartial`, isn't it?

Reply:
We should like that, but GHC does not infer it and the module would not compile if I remove this line.

Reply:
Ping me if you have further questions to this.